### PR TITLE
feat: add testdata to egctl/gwapi/xds lib

### PR DESCRIPTION
**Pull request by Xunzhuo**
**What type of PR is this?**

feat: add testdata to egctl/gwapi/xds lib

**What this PR does / why we need it**:

introduce `make testdata` to manage tons of json/yaml files of eg.

### CODE REVIEW

File: tools/make/golang.mk [hunk 1]
```diff
@@ -48,7 +48,10 @@ go.test.unit: ## Run go unit tests
 go.testdata.complete: ## Override test ouputdata
 	go test -timeout 30s -run ^TestTranslateXds github.com/envoyproxy/gateway/internal/xds/translator --override-testdata=true
 	go test -timeout 30s -run ^TestTranslate github.com/envoyproxy/gateway/internal/cmd/egctl --override-testdata=true
-
+	go test -timeout 30s -run ^TestExtractSubResourcesConfigDump github.com/envoyproxy/gateway/internal/cmd/egctl --override-testdata=true
```
Comment (arkodg) [inline tools/make/golang.mk:51]: this makes the target faster but also requires the developer to remember to add the Test name in here, can we just run this on all tests ?

File: internal/cmd/egctl/testdata/config/out/out.all.json [hunk 2]
```diff
@@ -1,1513 +1,2316 @@
 {
-	"default": {
```
Comment (arkodg) [inline internal/cmd/egctl/testdata/config/out/out.all.json:2]: why are these deleted

File: internal/cmd/egctl/config_test.go [hunk 3]
```diff
@@ -251,6 +260,18 @@ func readOutputConfig(filename string) ([]byte, error) {
 	return b, nil
 }
 
+func overrideOutputConfig(t *testing.T, data string, name ...string) {
```
[the remark this answers is not in the document]
Comment (Xunzhuo) [reply internal/cmd/egctl/config_test.go]: Done : )

File: tools/make/golang.mk (hunk 1 above)
Comment (Xunzhuo) [reply tools/make/golang.mk]: Yep, that is right.

File: internal/cmd/egctl/testdata/config/out/out.all.json (hunk 2 above)
Comment (Xunzhuo) [reply internal/cmd/egctl/testdata/config/out/out.all.json:2]: It is not deleted, just change the indent, we write it by hand before, now it is overidden by codes.

File: internal/ir/xds.go [hunk 4]
```diff
@@ -49,19 +49,19 @@ var (
 // +k8s:deepcopy-gen=true
 type Xds struct {
 	// AccessLog configuration for the gateway.
-	AccessLog *AccessLog
```
Comment (arkodg) [inline internal/ir/xds.go:52]: needs `omitempty` tags as well since they show up in the o/p as `null` atm

Comment (Xunzhuo) [reply internal/ir/xds.go:52]: Done

**Comment by Xunzhuo** [comment]
Some additional Updates:

1. Set fixed `LastTransitionTime` with the first commit time, otherwise each time we run `make testdata` will update over 100 files the `LastTransitionTime` with the current time, the changes should not be expected:
```
commit b86296d61439d115b288f01979e82865c8621f44
Author: Matt Klein <mattklein123@gmail.com>
Date:   Mon Apr 11 16:34:39 2022 -0600

    Initial commit
```
2. Add `make testdata` to `gen-check` in CI, so it will check if PR changes IR and test configs are updated with `make testdata` correctly.

cc @arkodg

**Comment by arkodg** [comment]
@Xunzhuo thoughts on setting any timestamp to `null` instead which is easier to comprehend, debug and fix in the testdata by the user, in case they are handwriting test files instead of using `make testdata`

**Comment by arkodg** [comment]
also if we add it to CI, it might cause a failure when a user adds the right content, in wrong order (for fields)

**Comment by Xunzhuo** [comment]
> @Xunzhuo thoughts on setting any timestamp to `null` instead which is easier to comprehend, debug and fix in the testdata by the user, in case they are handwriting test files instead of using `make testdata` ?

Right, I have updated it :)

**Comment by Xunzhuo** [comment]
> also if we add it to CI, it might cause a failure when a user adds the right content (who is unaware of `make testdata`) in the testdata yaml, in wrong order (for fields)

I think we should always suggest devs using make testdata instead of handwriting. Handwriting codes is a nightmare for everyone I think, and even if content is correct, but handwriting cannot promise format or indent is correct in json/yaml.

**Review by arkodg — APPROVED** [review]
LGTM thanks @Xunzhuo !
this should increase velocity of development by automating adding testdata with `make testdata`

**Review by qicz — COMMENTED** [review]
LGTM
